### ISSUE TITLE
Update Scrambled Spreadsheets

### ISSUE THREAD

**Issue by smokyblisters** [xwiki-labs/cryptpad#547]
A recent update appears to have scrambled spreadsheet information. I see CryptPad is aware of the issue and has "deployed a fix", according to a Twitter post. But my spreadsheets are still scrambled. 

**When can users expect the issue to be resolved?** Do I need to do anything on my end (I "logout everywhere" and have cleared cookies, but that hasn't helped)? 

On my end, a spreadsheet will load correctly, then load again (spinning wheel). The second load scrambles tab names (reverting to a mix of current and old tab names from previous document versions) and information. Formats and data are moved both within a tab (down a row or several) and across tabs (appearing to retain proper cell location, albeit on the wrong tab). 

The spreadsheet described above loads mostly blank cells where formulas are written, but clicking on a cell to edit it displays the proper formula. The calculated value doesn't display. Other spreadsheets simply load blank / greyed out tabs (no cells).

**Comment by ansuz** [comment]
I've marked two related issues as duplicates and closed them so that I can respond in one place.

For some background information: CryptPad's spreadsheet editor embeds OnlyOffice, which in my opinion is the best browser-based spreadsheet editor available under an open-source license. Unfortunately, it was developed with the intent of being used in combination with a server that performs a number of tasks for its clients, a mode of operation is incompatible with our goals with regard to user privacy.

We've done our best not to modify OnlyOffice directly, since that makes it easier for us to merge their continued improvements to the software. We've removed all the functions of their server and tried to wrap their editor with our own code which intercepts messages which would normally be sent to their server. This allows us to encrypt messages before storing them and to perform some other tasks like preventing other users from editing cells while we're in the process of modifying them ourselves.

The downside to this is that the spreadsheet editor works completely differently than other parts of the platform, it doesn't benefit from features that our other editors share (like a history navigation mode), and that it takes a disproportionate amount of time compared to our other applications to provide a comparable level of support. Spreadsheets are a very powerful tool, but editing them in a collaborative setting introduces a large number of edge cases which have made them very difficult to test.

The bugs reported in this issue and the others that have referenced it are all related to element _ids_. Unlike our other editors, when a client creates a new element in a sheet (like a workbook, chart, textarea, or image) it assigns it an id but does not necessarily share that id with other editors. When multiple users create such elements at the same time, they can end up having different names for the same data.

The recent issues are all related to this in one way or another. In some cases a number of changes to different worksheets ended up being applied to the same worksheet due to an incorrect id assignment. In other cases a set of changes were applied to a worksheet that no longer existed. In yet another situation it was possible for sheets not to load at all due to a reference to a data set that did not exist.

The instances of document corruption that followed our 3.17.1 release were due to an incorrect migration script that was intended to repair the ids assigned to these different types of special elements. Once we received reports of corruption on Wednesday May 13th we deployed an update to correct the issue. The update prioritized data generated before the previous day's migration script, since there was relatively little data produced in the intermediary period.

We believe that sheets should be stable now, however, we'll be keeping a close eye on things for the foreseeable future. Going forward we hope to implement the same history functionality that is present in our other editors so that users can roll back to a functional state in the event of corrupted data.

We're sorry for any difficulty this has caused. Paying users that require additional support to recover corrupted data can contact us via CryptPad.fr's support ticket functionality.

**Comment by smokyblisters** [comment]
All of my spreadsheets are still scrambled / corrupted. I'll ask for help via a support ticket. Thanks!

**Comment by smokyblisters** [comment]
Btw, thanks for the thorough response. I'm a paid member, I support you guys, and I'll continue to do so despite the inconvenience this week. 

I should have backed up my CryptDrive more frequently, anyway. Lesson learned!

**Comment by Tbaut** [comment]
I can confirm that my cryptpad is still acting unexpectedly. It says changes are saved, but they aren't, it creates copies of tabs with some data missing, formula aren't executed somehow, it show a formula in the edition field but the cell remain empty.

**Comment by HDValentin** [comment]
I have this issue on an own installation of Cryptpad. 
How can my administrator help me to recover a broken spreadsheet?

**Comment by ldubost** [comment]
For information, I helped two cryptpad.fr users, and exporting the spreasheet to a .bin and reimporting it in a new spreadsheet has worked for these cases.

**Comment by ldubost** [comment]
Note: the export was made from the read-only view of the spreadsheet. I cannot confirm it would be ok from the editable view

**Comment by HDValentin** [comment]
Thank you.
When I export the bin file it has a size of 1 KB.
The content is "undefined"
Here is the link to the corrupted pad:
https://cryptpad.organise.earth/sheet/#/2/sheet/edit/m6Vd-Bi7SfNTScbsGgf3x-io/

**Comment by ldubost** [comment]
@HDValentin indeed it seems to not give any valid data. Apparently you instance is 3.16 so it could be a different problem that broke the sheet.
The only solution here might be to go find the checkpoints to get an older version.. Using debug we might be able to find the checkpoints: https://cryptpad.organise.earth/debug/#/2/file/p0yDbZb5-pYeMWvmuhZ2AYOU/ 
However I'm not competent enough from there to tell how to rollback manually versions on the server.

**Comment by Tbaut** [comment]
> For information, I helped two cryptpad.fr users, and exporting the spreasheet to a .bin and reimporting it in a new spreadsheet has worked for these cases.

Thanks, that worked for me as well.

**Comment by smokyblisters** [comment]
> For information, I helped two cryptpad.fr users, and exporting the spreasheet to a .bin and reimporting it in a new spreadsheet has worked for these cases.

Yes, that fixed all of my spreadsheets. I was even able to time the moment I hit the export button. If I hit the export button on the first load of a document (you have to be quick), then the .bin file imported into a new spreadsheet generally contains fewer errors / scrambled data. If I do it on the second load of a document (spinning wheel or later), then the .bin file imported into a new spreadsheet requires much more formatting. Either way, your data are not lost.

**Comment by cyphunk** [comment]
as @smokyblisters mentions above and @ldubost a few up, but a few notes:
- it is possible to recover a corrupted spreadsheet. @ldubost did this through exporting a read only sheet. I can confirm this works on an editable sheet also.
- similar or same to what @smokyblisters mentioned, I had to press the export button immediately and before the corruption version appears. (in my case I could for a split second see the non-corrupted version before it was changed to the corrupted version). Once the export dialog appears there is no rush from that point and I can choose .bin or xls and name the file.
